[FIX] improve HED documentation (#619)
bids-standard/bids-specification: src/99-appendices/03-hed.md

# Appendix III: Hierarchical Event Descriptors

Hierarchical Event Descriptors (HED) are a controlled vocabulary of terms describing
events in a machine-actionable form so that algorithms can use the information without

> manual recoding. HED was originally developed with EEG in mind, but is applicable to
> all behavioral experiments.

Reviewer: adding a line break after each finished sentence:

```suggestion
manual recoding.
HED was originally developed with EEG in mind, but is applicable to all behavioral experiments.
```

> Each level of a hierarchical tag is delimited with a forward slash (`/`). A HED string
> contains one or more HED tags separated by commas (`,`). Parentheses (brackets, `()`)
> group tags and enable specification of multiple items and their attributes in a
> single **HED string** (see section 2.4 in
> [HED Tagging Strategy Guide](http://www.hedtags.org/downloads/HED%20Tagging%20Strategy%20Guide.pdf)).

Reviewer: again: line breaks after finished sentences.

```suggestion
Each level of a hierarchical tag is delimited with a forward slash (`/`).
A HED string contains one or more HED tags separated by commas (`,`).
Parentheses (brackets, `()`) group tags and enable specification of multiple items
and their attributes in a single **HED string** (see section 2.4 in
[HED Tagging Strategy Guide](http://www.hedtags.org/downloads/HED%20Tagging%20Strategy%20Guide.pdf)).
```

For more information about HED and tools available to validate and match HED
strings, please visit [www.hedtags.org](http://www.hedtags.org). Since
dedicated fields already exist for the overall task classification in the
sidecar JSON files (`CogAtlasID` and `CogPOID`), HED tags from the `Paradigm`
HED subcategory should not be used to annotate events.

## Annotating each event

There are several ways to associate HED annotations with events within the BIDS
framework. The most direct way is to use the `HED` column of the `*_events.tsv`
file to annotate events.

Example: An `*_events.tsv` annotated using HED tags for individual events.

```Text
onset  duration  HED
1.1    n/a       Event/Category/Experimental stimulus, Event/Label/CrossFix,  Sensory presentation/Visual, Item/Object/2D Shape/Cross

> 1.3    n/a       Event/Category/Participant response, Event/Label/ButtonPress, Action/Button press

Reviewer: here you replaced the "tab" character `	` with a bunch of "space" characters. 

I know that we have both variants in the specficiation. The "tab" character style of formatting the TSV examples has the advantage that users can simply copy and paste it ... at the expense that sometimes the columns are not aligned.

I am not sure what kind of style we should enforce throughout the spec, but consistency would probably pay off. Opinions @tsalo @effigies ? I know that @yarikoptic always wants "tab" characters.

Author: Apologies on the line breaks---I'll try to remember for next time.  I ended up changing the tabs to spaces because I found the table unreadable with the columns not aligned.  This is problematic for more complex tables.

...
```

The direct approach requires that each line in the events file be
annotated. Since there are typically thousands of events in each experiment,

> this method of annotation is not convenient unless the annotations are
> automatically generated. Usually annotations that appear in the `HED` column
> are specific to each individual event. Information that is common to groups
> of events can be annotated by category. Numerical values associated with each
> event can be annotated by value type.  Annotating by category and by value
> greatly reduces the effort required to HED tag data and improves the clarity
> for data users.

Reviewer: ```suggestion
this method of annotation is not convenient unless the annotations are
automatically generated.
Usually annotations that appear in the `HED` column are specific to each individual event.
Information that is common to groups of events can be annotated by category.
Numerical values associated with each event can be annotated by value type. 
Annotating by category and by value greatly reduces the effort required to HED tag data
and improves the clarity for data users.
```

## Annotating events by categories

> In many experiments, the event instances fall into a much smaller number of
> categories, and often these categories are labeled with numerical codes or short names.
> This categorical information usually corresponds to one or more columns in `*_events.tsv`
> representing categorical values. Instead of tagging this information for each
> individual event, you can assign HED tags for each distinct categorical value
> in an accompanying `*_events.json` sidecar and allow the analysis tools to make
> the association with individual event instances during analysis.  The column
> name in the `*_events.tsv` identifies the type of categorical variable. The following
> `*_events.tsv` file has one categorical variable called `mycodes` that takes
> on three possible values: `Fixation`, `Button`, and `Target`.

Reviewer: line breaks:

```suggestion
In many experiments, the event instances fall into a much smaller number of
categories, and often these categories are labeled with numerical codes or short names.
This categorical information usually corresponds to one or more columns in `*_events.tsv`
representing categorical values.
Instead of tagging this information for each individual event,
you can assign HED tags for each distinct categorical value
in an accompanying `*_events.json` sidecar and allow the analysis tools to make
the association with individual event instances during analysis.
The column name in the `*_events.tsv` identifies the type of categorical variable.
The following `*_events.tsv` file has one categorical variable called `mycodes` that takes
on three possible values: `Fixation`, `Button`, and `Target`.
```

Example: An `*_events.tsv` containing the `mycodes` categorical column.

```Text
onset  duration  mycodes
1.1    n/a       Fixation
1.3    n/a       Button

> 1.8    n/a       Target

Reviewer: same comment as above --> not sure if in the whole BIDS spec, we should format the TSV examples using "tab" characters ... or a bunch of spaces.

tab seems cleaner and more consistent (it's always one character, ... not a varying number of space characters), but space characters currently make the examples look nicer :shrug:

       },
       "HED": {
          "go": "Event/Category/Experimental stimulus, Event/Label/RedSquare,
          Event/Description/A red square is displayed to indicate starting,
          Sensory presentation/Visual, Item/Object/2D Shape/Square,
          Attribute/Visual/Color/Red, Attribute/Visual/Rendering type/Screen,
          Attribute/Location/Screen/Center",
          "stop": "Event/Category/Experimental stimulus, Event/Label/BlueSquare,
          Event/Description/A blue square is displayed to indicate stopping,
          Sensory presentation/Visual, Item/Object/2D Shape/Square,
          Attribute/Visual/Color/Blue, Attribute/Visual/Rendering type/Screen,
          Attribute/Location/Screen/Center",
       }
   },
   "response_time": {
       "LongName": "Response time after stimulus",
       "Description": "Time from stimulus presentation until subject presses button",
       "Units": "ms",
       "HED": "Attribute/Response start delay/# ms, Action/Button press"
   },
   "stim_file": {
       "LongName": "Stimulus file name",
       "Description": "Relative path of the stimulus image file",
       "HED": "Attribute/File/#"
   }
}
```

## Best practices

> Most studies will have event categorical variables and value variables that are common across many
> of the datasets in the study. You should try to annotate these columns in a `*_events.json` sidecar
> as high in the study hierarchy as possible to avoid duplicate annotations. Annotations that can be
> placed in sidecars are preferred to those placed directly in the HED column, because they are simpler,
> more compact, and less prone to inconsistent annotation. Downstream tools should not distinguish
> between tags specified using the explicit HED column and the categorical specifications, but should
> form the union before analysis. Further, the [inheritance principle](../02-common-principles.md#the-inheritance-principle)
> applies, so the data dictionaries can appear higher in the BIDS hierarchy.

Reviewer: line breaks:

```suggestion
Most studies will have event categorical variables and value variables that are common across many
of the datasets in the study.
You should try to annotate these columns in a `*_events.json` sidecar
as high in the study hierarchy as possible to avoid duplicate annotations.
Annotations that can be placed in sidecars are preferred to those placed directly in the HED column,
because they are simpler, more compact, and less prone to inconsistent annotation.
Downstream tools should not distinguish between tags specified using the explicit HED column and
the categorical specifications, but should form the union before analysis.
Further, the [inheritance principle](../02-common-principles.md#the-inheritance-principle) applies,
so the data dictionaries can appear higher in the BIDS hierarchy.
```

> You can should try to annotate in as much detail as possible. The HED path structure makes it easy
> for analysis tools to extract tags at different levels of detail, for example only consider 2D
> shapes for stimuli, but not worry about color or shape details.

Reviewer: corrected a typo ("can should"), added line breaks.

I think these two sentences (1) and (2) also need a more explicit logical connection --> more like: `(1) try to tag well ... because then (2)` ... currently it's just `(1) ... (2)`, without explicit connection

```suggestion
You should try to annotate your data in as much detail as possible.
The HED path structure makes it easy for analysis tools to extract tags at different levels of detail,
for example only consider 2D shapes for stimuli, but not worry about color or shape details.
```
